docs/control_tuple_support.md

```diff
@@ -0,0 +1,207 @@
+# Custom Control Tuple (User Defined Control Tuple) Support
+
+## Introduction
+
+Custom control tuple support in Apache Apex gives the user the capability to insert user defined control tuples in the data flow. For analogy, the engine already supports few pre-defined control tuples like BEGIN_WINDOW, END_WINDOW, etc. Until now, we did not have the support for applications to insert their own control tuples.
+
+## Terminology
+All discussion in this document is related to Control Tuples generated by user defined logic. The document may refer to these tuples as *Control Tuples*, *User Defined Control Tuples* or *Custom Control Tuples* interchangeably.
+
+### Definition
+A user defined control tuple could be any user defined object which implements a ControlTuple interface.
+
+See [Delivery Semantics](#delivery-semantics) for details on DeliveryType
+```
+public interface ControlTuple
+{
+  DeliveryType getDeliveryType();
+
+  enum DeliveryType
+  {
+    IMMEDIATE,
+    END_WINDOW
+  }
+}
+
+```
+
+Example user defined control tuple:
+```
+public class TestControlTuple implements ControlTuple
+{
+  public long data;
+  public boolean immediate;
+
+  // For Kryo
+  public TestControlTuple()
+  {
+    data = 0;
+    immediate = false;
+  }
+
+  // Constructor
+  public TestControlTuple(long data, boolean immediate)
+  {
+    this.data = data;
+    this.immediate = immediate;
+  }
+
+  @Override
+  public DeliveryType getDeliveryType()
+  {
+    if (immediate) {
+      return DeliveryType.IMMEDIATE;
+    } else {
+      return DeliveryType.END_WINDOW;
+    }
+  }
+}
+```
+
+
+## Use cases
+A control tuple may be used by an application to trigger some sort of action in a remote operator. For example, the source operator might want to notify the last operator that it has emitted all the data in a file and that the file has now ended. Let's call this an *End-Of-File* control tuple. Once the last operator gets the *End-Of-File* tuple, it would, say, close the destination file it was writing and create a new file.
+
+More use cases which were discussed during the requirements of this feature are as follows:
+
+1. **Batch support** - We need to tell all operators of the physical DAG when a
+batch starts and ends, so the operators can do whatever that is needed upon
+the start or the end of a batch.
+2. **Watermark** - To support the concepts of event time windowing, the
+watermark control tuple is needed to tell which windows should be
+considered late.
+3. **Changing operator properties** - We do have the support of changing
+operator properties on the fly, but with a custom control tuple, the
+command to change operator properties can be window aligned for all
+partitions and also across the DAG.
```

Comment:
This is hard to follow and needs more explanation; for instance, are you talking about changing different properties in different operators all with one control tuples ? Also, the fact that these tuples are sent, in contrast to data tuples, to _all_ partitions is worth pointing out.